### PR TITLE
Chatglm2/3 bf16 pipeline support

### PR DESCRIPTION
**Pull request by a3213105**
First version of Chatglm2/3 bf16 whole pipeline support.

### CODE REVIEW

File: src/layers/rotary_embedding_chatglm2.cpp [hunk 1]
```diff
@@ -87,65 +87,112 @@ void ChatGLM2RotaryEmbedding::glm2CalEmb() {
...
                 int pos = position_ids[seq];
                 float *pcos = emb_cos + pos * dim;
                 float *psin = emb_sin + pos * dim;
-
-#pragma omp simd
-                for (int i = 0; i < half; i += 2) {
-                    auto t1 = p1[i];
-                    p1[i] = p1[i] * pcos[i] - p1[i + 1] * psin[i];
-                    p1[i + 1] = p1[i + 1] * pcos[i] + t1 * psin[i];
+                for (int i = 0; i < dim; i += 32) {
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.cpp:170]:
```c++
#include <immintrin.h>
#include <stdio.h>

void interleave_merge(__m512 a, __m512 b, __m512 *result) {
    // 生成掩码
    //__m512i mask = _mm512_set_epi32(0x17, 0x07, 0x16, 0x06, 0x15, 0x05, 0x14, 0x04, 0x13, 0x03, 0x12, 0x02, 0x11, 0x01, 0x10, 0x00);
    __m512i mask = _mm512_set_epi32(0x1f, 0x0f, 0x1e, 0x0e, 0x1d, 0x0d, 0x1c, 0x0c, 0x1b, 0x0b, 0x1a, 0x0a, 0x19, 0x09, 0x18, 0x08);

    // 交叉合并数据
    __m512 merged = _mm512_permutex2var_ps(a, mask, b);

    // 存储结果
    *result = merged;
}

int main() {
    // 假设有两个__m512寄存器 a 和 b
    __m512 a = _mm512_set_ps(8.0, 7.0, 6.0, 5.0, 4.0, 3.0, 2.0, 1.0, 16.0, 15.0, 14.0, 13.0, 12.0, 11.0, 10.0, 9.0);
    __m512 b = _mm512_set_ps(24.0, 23.0, 22.0, 21.0, 20.0, 19.0, 18.0, 17.0, 32.0, 31.0, 30.0, 29.0, 28.0, 27.0, 26.0, 25.0);

    // 进行交叉合并
    __m512 result;
    interleave_merge(a, b, &result);

    // 输出结果
    float* res = (float*)&result;
    for (int i = 0; i < 16; ++i) {
        printf("%f ", res[i]);
    }
    printf("\n");

    return 0;
}
```

$ g++ test.cc -march=native
$ ./a.out
1.000000 17.000000 2.000000 18.000000 3.000000 19.000000 4.000000 20.000000 5.000000 21.000000 6.000000 22.000000 7.000000 23.000000 8.000000 24.000000

File: src/layers/rotary_embedding_chatglm2.cpp [hunk 2]
```diff
@@ -87,65 +87,112 @@ void ChatGLM2RotaryEmbedding::glm2CalEmb() {
 //     x_out2 = x_out2.flatten(3)
 //     return torch.cat((x_out2, x_pass), dim=-1)
 
-void ChatGLM2RotaryEmbedding::forward(float *buf, int bufStride, int batch_size, int seq_len, int qk_size,
-        int hidden_size_per_attention_head, const int *position_ids) {
-    int dim = inv_freq_size * 2;
-    REQUIRES(dim == hidden_size_per_attention_head, "Incorrect shape, last dimention is not the head size.");
+// void ChatGLM2RotaryEmbedding::forward(float *buf, int bufStride, int batch_size, int seq_len, int qk_size,
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.cpp:90]: Delete those

File: src/layers/rotary_embedding_chatglm2.h [hunk 3]
```diff
@@ -35,10 +36,12 @@ class ChatGLM2RotaryEmbedding {
 
     ~ChatGLM2RotaryEmbedding() {}
 
-    void forward(float *buf, int bufStride, int batch_size, int seq_len, int qk_size,
-            int hidden_size_per_attention_head, const int *position_ids);
+    // void forward(float *buf, int bufStride, int batch_size, int seq_len, int qk_size,
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.h:39]: delete those.

File: src/models/chatglm2.cpp [hunk 4]
```diff
@@ -21,8 +21,13 @@
...
 template <typename WeiT>
 ChatGLM2<WeiT>::ChatGLM2(const std::string &modelPath, const std::string &modelType)
-    : CommonDecoder<Attention<WeiT, ChatGLM2RotaryEmbedding, RmsNorm, float, float, float, true>,
-            ChatGLM2MLP<WeiT, float, float, float, RmsNorm, true>>(modelPath, modelType) {
+    : CommonDecoder<Attention<WeiT, ChatGLM2RotaryEmbedding, RmsNorm, typename TypeSelector<WeiT>::InType,
+                            typename TypeSelector<WeiT>::ImType, typename TypeSelector<WeiT>::OutType, true>,
+            ChatGLM2MLP<WeiT, typename TypeSelector<WeiT>::InType, typename TypeSelector<WeiT>::ImType,
+                    typename TypeSelector<WeiT>::OutType, RmsNorm, true>,
+            typename TypeSelector<WeiT>::KVCacheType>(modelPath, modelType) {
+    // : CommonDecoder<Attention<WeiT, ChatGLM2RotaryEmbedding, RmsNorm, float, float, float, true>,
```
Comment (changqi1) [inline src/models/chatglm2.cpp:29]: delete those.

File: src/models/chatglm2.h [hunk 5]
```diff
@@ -23,15 +23,23 @@
...
 template <typename WeiT>
-class ChatGLM2 : public CommonDecoder<Attention<WeiT, ChatGLM2RotaryEmbedding, RmsNorm, float, float, float, true>,
-                         ChatGLM2MLP<WeiT, float, float, float, RmsNorm, true>> {
+class ChatGLM2
+    : public CommonDecoder<Attention<WeiT, ChatGLM2RotaryEmbedding, RmsNorm, typename TypeSelector<WeiT>::InType,
+                                   typename TypeSelector<WeiT>::ImType, typename TypeSelector<WeiT>::OutType, true>,
+              ChatGLM2MLP<WeiT, typename TypeSelector<WeiT>::InType, typename TypeSelector<WeiT>::ImType,
+                      typename TypeSelector<WeiT>::OutType, RmsNorm, true>,
+              typename TypeSelector<WeiT>::KVCacheType> {
+    // : public CommonDecoder<Attention<WeiT, ChatGLM2RotaryEmbedding, RmsNorm, float, float, float, true>,
```
Comment (changqi1) [inline src/models/chatglm2.h:32]: delete those.

File: src/layers/rotary_embedding_chatglm2.cpp (hunk 1 above)
Comment (a3213105) [reply src/layers/rotary_embedding_chatglm2.cpp:170]: thanks

File: src/layers/rotary_embedding_chatglm2.cpp [hunk 6]
```diff
@@ -138,17 +108,52 @@ void ChatGLM2RotaryEmbedding::forward(
...
+void prepare_sincos(__m512 a, __m512 b, __m512 *result) {
+    __m512i mask = _mm512_set_epi32(0x1e, 0x1c, 0x1a, 0x18, 0x16, 0x14, 0x12, 0x10,
+                                    0x0e, 0x0c, 0x0a, 0x08, 0x06, 0x04, 0x02, 0x00);
+
+    __m512 merged = _mm512_permutex2var_ps(a, mask, b);
+    *result = merged;
+}
+
+void load_qk(__m512 a, __m512 b, __m512 *result0, __m512 *result1) {
+    __m512i mask0 = _mm512_set_epi32(0x1e, 0x1c, 0x1a, 0x18, 0x16, 0x14, 0x12, 0x10,
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.cpp]: const __m512i

File: src/layers/rotary_embedding_chatglm2.cpp [hunk 7]
```diff
@@ -138,17 +108,52 @@ void ChatGLM2RotaryEmbedding::forward(
...
+            0x0e, 0x0c, 0x0a, 0x08, 0x06, 0x04, 0x02, 0x00);
+
+    __m512i mask1 = _mm512_set_epi32(0x1f, 0x1d, 0x1b, 0x19, 0x17, 0x15, 0x13, 0x11,
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.cpp]: const __m512i

File: src/layers/rotary_embedding_chatglm2.cpp [hunk 8]
```diff
@@ -138,17 +108,52 @@ void ChatGLM2RotaryEmbedding::forward(
...
+            0x0f, 0x0d, 0x0b, 0x09, 0x07, 0x05, 0x03, 0x01);
+
+    __m512 merged = _mm512_permutex2var_ps(a, mask0, b);
+    *result0 = merged;
+    merged = _mm512_permutex2var_ps(a, mask1, b);
+    *result1 = merged;
+}
+
+void store_qk(__m512 a, __m512 b, __m512 *result0, __m512 *result1) {
+    __m512i mask0 = _mm512_set_epi32(
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.cpp:144]: const __m512i

File: src/layers/rotary_embedding_chatglm2.cpp [hunk 9]
```diff
@@ -138,17 +108,52 @@ void ChatGLM2RotaryEmbedding::forward(
...
+            0x17, 0x07, 0x16, 0x06, 0x15, 0x05, 0x14, 0x04, 0x13, 0x03, 0x12, 0x02, 0x11, 0x01, 0x10, 0x00);
+
+    __m512i mask1 = _mm512_set_epi32(
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.cpp:147]: const __m512i

File: src/layers/rotary_embedding_chatglm2.cpp [hunk 10]
```diff
@@ -138,17 +108,52 @@ void ChatGLM2RotaryEmbedding::forward(
...
+}
+
+void load_qk(__m512 a, __m512 b, __m512 *result0, __m512 *result1) {
+    __m512i mask0 = _mm512_set_epi32(0x1e, 0x1c, 0x1a, 0x18, 0x16, 0x14, 0x12, 0x10,
+            0x0e, 0x0c, 0x0a, 0x08, 0x06, 0x04, 0x02, 0x00);
+
+    __m512i mask1 = _mm512_set_epi32(0x1f, 0x1d, 0x1b, 0x19, 0x17, 0x15, 0x13, 0x11,
+            0x0f, 0x0d, 0x0b, 0x09, 0x07, 0x05, 0x03, 0x01);
+
+    __m512 merged = _mm512_permutex2var_ps(a, mask0, b);
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.cpp:137]: *result0  = _mm512_permutex2var_ps(a, mask0, b);  Why not?

File: src/layers/rotary_embedding_chatglm2.cpp [hunk 11]
```diff
@@ -138,17 +108,52 @@ void ChatGLM2RotaryEmbedding::forward(
...
+                    p1[i + 1] = p1[i + 1] * pcos[i] + t1 * psin[i];
                 }
                 off += qStride;
             }
         }
     }
 }
 
+void prepare_sincos(__m512 a, __m512 b, __m512 *result) {
+    __m512i mask = _mm512_set_epi32(0x1e, 0x1c, 0x1a, 0x18, 0x16, 0x14, 0x12, 0x10,
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.cpp]: const __m512i

**Review by abenmao — APPROVED** [review]

File: src/layers/rotary_embedding_chatglm2.cpp [hunk 12]
```diff
@@ -111,41 +113,87 @@ void ChatGLM2RotaryEmbedding::forward(float *buf, int bufStride, int batch_size,
...
     int dim = inv_freq_size * 2;
     REQUIRES(dim == qk_shape[3], "Incorrect shape, last dimention is not the head size.");
     const int batch_size = qk_shape[0];
     const int seq_len = qk_shape[1];
-    const int head_num = qk_shape[2] +  qk_shape[4];
+    const int head_num = qk_shape[2] + qk_shape[4];
     const int half = inv_freq_size;
-
-#pragma omp parallel for
+    // #pragma omp parallel for
```
Comment (changqi1) [inline src/layers/rotary_embedding_chatglm2.cpp:159]: Why not to use it to parallel for?

Comment (a3213105) [reply src/layers/rotary_embedding_chatglm2.cpp:159]: for the same parameters：
2 sockets:
without  #pragma omp parallel for
Througput:      12.+ tokens/s

with #pragma omp parallel for
Througput:      11.+ tokens/s

we can assume that no benefit gets from OMP PARALLEL.

Comment (a3213105) [reply src/layers/rotary_embedding_chatglm2.cpp]: maybe we are better to align with other rotary embedding implement.

**Review by changqi1 — APPROVED** [review]

**Pull request merged**